File: 3691-maximum-total-subarray-value-ii/3691-maximum-total-subarray-value-ii.cpp
#include<bits/stdc++.h>
using namespace std;

#define mod ((int)1e9+7)
#define ll long long

#define vi vector<int>
#define vii vector<vi>
#define pi pair<ll,ll>
#define vpi vector<pi>

#define F first
#define S second
#define pb push_back
#define bp pop_back
#define mp make_pair
#define all(x) x.begin(),x.end()



#define F first
#define S second
#define pb push_back
#define bp pop_back
#define mp make_pair
#define all(x) x.begin(),x.end()

class Solution {
public:

    static const int K = 25;
    vector<vpi> st;

    pi merge(pi& a ,pi& b){
        return {max(a.F,b.F),min(a.S,b.S)};
    }

    void build(int n ,vi& nums){

        st.resize(K+1,vpi(n+1));

        for(int j=0;j<n;j++)
            st[0][j] = {nums[j],nums[j]};

        for(int i=1;i<=K;i++){
            for(int j=0;j + (1 << (i-1)) <= n;j++){
                st[i][j] = merge(st[i-1][j],st[i-1][j + (1 << (i - 1))]);
            }
        }
    }

    pi rangeQuery(int l ,int r){
        pi sol = {-1e18,1e18};
        for(int i=K;i>=0;i--){
            if(r - l + 1 >= (1 << i)){
                sol = merge(sol,st[i][l]);
                l += (1 << i);
            }
        }

        return sol;
    } 

    ll maxTotalValue(vi& nums, int k) {

        int n = nums.size();
        build(n,nums);
        priority_queue<pair<ll,pair<int,int>>> pq;

        for(int l=0;l<n;l++){
            pi p = rangeQuery(l,n-1);
            ll val = p.F - p.S;
            pq.push({val,{l,n-1}});
        }

        ll sol = 0;
        while(k--){
            pi range = pq.top().S;
            ll val = pq.top().F;
            pq.pop();
            sol += val;

            int l = range.F;
            int r = range.S; 
            if(r-1 >= l){
                pi p = rangeQuery(l,r-1);
                ll val = p.F - p.S;
                pq.push({val,{l,r-1}});
            }
        }

        return sol;
    }
};

File: 3691-maximum-total-subarray-value-ii/3691-maximum-total-subarray-value-ii.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;

public class Main{
}

class SparseTable{
    static int K = 25;
    int[][] min ,max;

    SparseTable(int[] nums){
        int n = nums.length;
        min = new int[K+1][n+1];
        max = new int[K+1][n+1];
        for(int i=0;i<n;i++){
            min[0][i] = nums[i];
            max[0][i] = nums[i];
        }

        for(int i=1;i<=K;i++){
            for(int j=0;j + (1 << (i-1)) <= n;j++){
                min[i][j] = min(min[i-1][j],min[i-1][j + (1 << (i-1))]);
                max[i][j] = max(max[i-1][j],max[i-1][j + (1 << (i-1))]);
            }
        }

    }

    int rangeQuery(int l ,int r){

        int mn = (int)1e10;
        int mx = -(int)1e10;

        for(int i=K;i>=0;i--){
            int ln = r - l + 1;
            if(ln >= (1 << i)){
                mn = min(mn,min[i][l]);
                mx = max(mx,max[i][l]);
                l += (1 << i);
            }
        }

        return mx - mn;
    }

    int min(int a ,int b){
        if(a < b)
            return a;
        return b;
    }

    int max(int a ,int b){
        if(a > b)
            return a;
        return b;
    }
}


class Solution {



    public long maxTotalValue(int[] nums, int k) {

        // l ,n - 1 its the maximum value hold , and the value was monotonically increasing 

        int n = nums.length;
        SparseTable st = new SparseTable(nums);

        PriorityQueue<int[]> pq = new PriorityQueue<>((a,b) -> b[0] - a[0]);

        for(int l=0;l<n;l++){
            int x = st.rangeQuery(l,n-1);
            int[] tmp = {x,l,n-1};
            pq.add(tmp);
        }

        long sol = 0;
        while(k-- > 0){
            int[] tmp = pq.poll();
            sol += tmp[0];
            int l = tmp[1] ,r = tmp[2];
            if(r-1 >= l){
                tmp[0] = st.rangeQuery(l,r-1);
                tmp[1] = l;
                tmp[2] = r-1;
                pq.add(tmp);
            }
        }

        return sol;
    }
}